### PR TITLE
Focus on egglog for optimization, remove the greedy

### PR DESCRIPTION
**Pull request by kvark**
Consolidates the remaining inference work on main `4a66a27`, replacing #193 and #196–199 (#194/#195 came through #193). Includes the GEMV/layout, projection, pipeline-preparation, readback and cooperative-correctness changes already reviewed here.

### Construction-time search (`b06629c`)

- One bounded egglog rewrite engine is now the default; the separately implemented greedy engine is removed. Legacy environment overrides are mapped explicitly.
- Keep bounded equivalent graph forms until their implementations have been tuned and compared. Reuse the normal compiler, allocator and kernel probes.
- `train::build_measured` takes initialization and read-only qualification callbacks. It checks outputs, gradients and persistent updates against the caller's numerical contract, restoring private state between runs and before return.
- Attention splits lower before allocation. The live attention installer and submission tuner are deleted; there is no buffer-growth/alias/profiling-index repair on a running application.
- Reuse qualified kernel comparisons within one build, not whole-program timings or validation. Reject invalid challengers; an invalid incumbent aborts construction.
- Precision boundaries remain opaque to rewrites. The existing integration suite caught a forward/derivative producer-identity issue in cross-entropy; it is fixed without changing tolerances.

[Design, limits and research comparison](https://github.com/kvark/meganeura/blob/b06629cb19467ac00e75f3b383bd3ef70442d2a7/docs/compiler-search.md).

This removes roughly **370 lines of non-test compiler/runtime/shader source**, but the migration plus recorded measurements is **+327 net lines**, including documentation, the benchmark adapter and broad tests. The improvement is architectural, not a dramatic total-LOC reduction.

Large-graph alternative extraction currently explores one verified repeated region; stateful/mixed-precision regions can be skipped explicitly. Uniform attention splits, finite budgets and representative calibration inputs limit the search. There is no persistent measured-plan cache.

No reusable command recordings, new shader families, dependencies, test executables or binary/raw benchmark artifacts. No new cohort was collected.

### Review follow-up (`f3fe2f2`, `6a5e9c3`)

- Addressed all four inline comments: explicit postorder traversal, directly interned node-ID literals, one operator match, allocation-free exclusion lookup, shared immutable exclusion sets, and term-ID candidate deduplication.
- Ordinary extraction now shares one extractor/cost table and term DAG across roots of an unchanged segment. The segment report replaces eleven separate arguments. This follow-up removes 43 production lines; including the opt-in CPU timing probe, `src/` is six lines smaller.
- CPU-only warm training-preparation medians (including differentiation/copies): SmolLM2 **734→92 ms**, SmolVLA **387→52 ms**, diffusion **521→88 ms**. The separate bounded alternative enumerator improves roughly 11–15%. No GPU-speedup claim. [Methods and full results](https://github.com/kvark/meganeura/blob/f3fe2f26edb14d655ed4a5569466ca5918149548/docs/compiler-search.md#cpu-overhead-and-review-follow-up).
- Paper-only commit separates RPL-U qualification from paired GPU timing. All CPU/GPU timing ratios and preparation comparisons are removed, derived counts regenerated, and raw archives preserved. Mendocino is pending, with [exact qualification instructions](https://github.com/kvark/meganeura/blob/6a5e9c3a7863692a6e0cb91e3c85503fe2b40d80/paper/p3hpc/QUALIFICATION.md).
- **Next cohort is not launch-ready just by merging/updating the pin.** Inferena still uses `build` then kernel tuning before input initialization. It needs an initialized/qualified `build_measured` adapter, search receipts/budgets, API migration and a five-model/two-contract NVIDIA+Intel pilot. Prior GGUF calibration variability and preparation costs below remain open.

### Verification

- NVIDIA: all-feature unit/integration suite **789 passed, 22 ignores** (one additional opt-in CPU timing probe), without the CI backprop/bit-exact environment skips.
- Rust host coverage: **87.15% lines, 85.19% functions, 86.53% regions**. WGSL is not instrumented; these local hardware results are not directly comparable with lavapipe CI coverage.
- Intel: **465 library tests** including opt-in GPU checks, plus the attention and gradient suites; whole-graph egglog cross-entropy is checked too.
- No-feature all-target checks, all-feature Clippy, formatting and documentation pass.
- Prior migration [CI](https://github.com/kvark/meganeura/actions/runs/35535427014) passed. Fresh CI for this follow-up is pending. Known Naga Workgroup ArrayStride diagnostics remain unchanged.
- Final-cohort archives re-audit successfully; the paper rebuilds as 13 pages without overflow or unresolved references. Six legacy artifact tests pass.

### Matched GGUF regression check

Exact `789bcb6` versus source commit `b06629c`, two fresh processes per arm/GPU with the second pair reversed. Same fixed SmolLM2-135M F16 GGUF workload. [Results, reproduction and limitations](https://github.com/kvark/meganeura/blob/dd743f4/bench/gguf_latency.md#calibrated-construction-checkpoint-2026-09-20).

Median of process medians; execution in milliseconds, preparation in seconds:

| GPU | Prefill before / after | Decode before / after | Preparation before / after |
| --- | ---: | ---: | ---: |
| RTX 5070 | 7.251 / 7.404 | 1.425 / 1.463 | 15.40 / 37.77 |
| Arc B570 | 14.060 / 14.062 | 3.871 / 3.786 | 31.75 / 63.76 |

NVIDIA is 2–3% slower in these aggregate medians. Intel decode varies from 3.541 to 4.032 ms after the change, so its apparent aggregate gain is not stable. **No performance-neutrality claim.** Preparation is costlier and uses a different budget: complete construction/qualification now counts, unlike the old kernel-only allowance.

All eight saved output sets are finite and preserve all 33 independent CPU-reference token choices; maximum per-row relative L2 is 1.33e-5 / 1.77e-4. The full-output gate rejects some NVIDIA prefill choices; it was not relaxed. This GGUF graph's stateful outlined region is explicitly skipped by logical alternative extraction, so these timings do not demonstrate a graph-alternative win.

Follow-up priorities are candidate-construction reuse, calibration stability and pure-region extraction within larger stateful graphs. The new kernel research remains separate.

### Separate research

The [source-only experiment](https://github.com/kvark/meganeura/tree/experiment/egglog-measured-2026-09-20) retains the dense split-K, serial-reduction and other kernel studies. Those new kernel families are not pulled into this refactoring. The submitted paper cohort remains unchanged.

### CODE REVIEW

File: src/runtime/tuning/attention.rs [hunk 1]
```diff
@@ -0,0 +1,994 @@
...
+        combine.label = label(&combine.shader);
+        combine.input_buffers = vec![partials];
+        combine.output_buffer = output;
+        combine.workgroups = [self.block_len, self.num_heads, 1];
+        vec![first, combine]
+    }
+}
+
+fn plain(d: &Dispatch) -> bool {
+    !d.use_coop
```
Comment (kvark) [inline src/runtime/tuning/attention.rs]: this approach is not scale-able. We could add another field to Dispatch and this condition wouldn't even no. At the very least, it should pattern match

File: src/runtime/tuning/attention.rs [hunk 2]
```diff
@@ -0,0 +1,994 @@
...
+            bindings.push(d.output_buffer);
+        }
+        let physical: Vec<_> = bindings
+            .iter()
+            .map(|b| session.alias.map[b.0 as usize])
+            .collect();
+        if physical
+            .iter()
+            .enumerate()
+            .any(|(i, p)| physical[..i].contains(p))
```
Comment (kvark) [inline src/runtime/tuning/attention.rs]: this is quadratic complexity? could we instead just sort and then dedup?

File: src/runtime/tuning/attention.rs [hunk 3]
```diff
@@ -0,0 +1,994 @@
...
+                if variant == 0 {
+                    references.push(output);
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+fn qualify_reference(
```
Comment (kvark) [inline src/runtime/tuning/attention.rs]: This is all very complex. Can't there be a more general way to say "ok, you have a graph, now this specific parameter needs to have the following tuning range". I don't want to see that custom thousand-line boilerplate specific to each op.

File: src/runtime/tuning/attention.rs (hunk 1 above)
Comment (kvark) [reply src/runtime/tuning/attention.rs]: Changed eligibility to an exhaustive Dispatch pattern (3829907). The follow-up Kernel enum replaces nine independent implementation flags/options with one variant, and the pattern accepts only Kernel::Default. Adding a Dispatch field now requires revisiting the match; adding a kernel kind does not silently opt it into this probe.

File: src/runtime/tuning/attention.rs (hunk 2 above)
Comment (kvark) [reply src/runtime/tuning/attention.rs]: Changed to sort_unstable/dedup on a clone of the physical binding list (3829907), preserving the original binding order for placement bookkeeping.

File: src/optimize/search.rs [hunk 4]
```diff
@@ -0,0 +1,556 @@
...
+        let value = match *terms.get(id) {
+            Term::App(ref head, ref args) => {
+                let inputs = args
+                    .iter()
+                    .map(|&arg| visit(egraph, terms, arg, values, edges))
+                    .collect::<Result<Vec<_>, _>>()?;
+                let value = egraph
+                    .lookup_function(head, &inputs)
+                    .ok_or("extracted term is missing from the e-graph")?;
+                if head != "Leaf" {
```
Comment (kvark) [inline src/optimize/search.rs:122]: why is graph traversal using any string operations?

File: src/optimize/search.rs [hunk 5]
```diff
@@ -0,0 +1,556 @@
...
+                    .eval_expr(&terms.term_to_expr(&id, egglog::ast::Span::Panic))
+                    .map_err(|e| e.to_string())?
+                    .1
+            }
+        };
+        values.insert(id, value);
+        Ok(value)
+    }
+    let mut result = Vec::new();
+    visit(egraph, terms, root, &mut HashMap::new(), &mut result)?;
```
Comment (kvark) [inline src/optimize/search.rs]: should this be rewritten with an explicit DFS stack instead of recursion?

File: src/optimize/search.rs [hunk 6]
```diff
@@ -0,0 +1,556 @@
...
+        return Err("region has no observable output".into());
+    }
+    let full_precision = graph.node(roots[0] as u32).requires_full_precision;
+    for id in segment
+        .shifts
+        .iter()
+        .flat_map(|shift| segment.ids.iter().map(move |id| id + shift))
+    {
+        let node = &graph.nodes()[id];
+        if matches!(
```
Comment (kvark) [inline src/optimize/search.rs]: You are over-using `matches!` macro for no good reason. Our CONTRIBUTING md style guide should already say to prefer simpler constructs. In this case, a single `match` statement would be nicer than those 2 different `matches!` invocations.

File: src/optimize/search.rs [hunk 7]
```diff
@@ -0,0 +1,556 @@
...
+    };
+    let ids: HashSet<_> = segment.ids.iter().copied().collect();
+    let uses = super::external_uses(graph, &segment);
+    let ext_maps = segment
+        .shifts
+        .iter()
+        .map(|&shift| super::instance_ext_map(graph, &segment, &uses, shift))
+        .collect::<Option<Vec<_>>>()
+        .ok_or("region instances have ambiguous external edges")?;
+    let mut pending = VecDeque::from([Vec::new()]);
```
Comment (kvark) [inline src/optimize/search.rs]: is this why egglog is much slower? this search is riddled with allocations

File: src/optimize/search.rs (hunk 4 above)
Comment (kvark) [reply src/optimize/search.rs:122]: Addressed in f3fe2f2. Traversal no longer constructs/evaluates an AST to recover node-ID literals: it interns the i64 directly. Candidate deduplication now uses TermId in one shared DAG, not formatted expression strings. Egglog itself exposes String constructor names and name-based lookup, so those remain borrowed at that API boundary; I did not add a duplicate local operator registry.

File: src/optimize/search.rs (hunk 5 above)
Comment (kvark) [reply src/optimize/search.rs]: Replaced the recursive visitor with an explicit postorder stack and a TermId-indexed value vector in f3fe2f2. Shared subterms are visited once. Existing nested/shared-root and repeated-region tests pass.

File: src/optimize/search.rs (hunk 6 above)
Comment (kvark) [reply src/optimize/search.rs]: Changed to one match: leaves continue, mutating operators reject the region, and the remaining operators proceed to precision/arity checks (f3fe2f2).

File: src/optimize/search.rs (hunk 7 above)
Comment (kvark) [reply src/optimize/search.rs]: Yes, there was unnecessary allocation here. f3fe2f2 shares immutable exclusion sets between the queue, visited set and extractor; the cost callback now compares borrowed names/argument slices without allocating an Edge. The larger cost was ordinary extraction recomputing its whole cost table per root: sharing one extractor per segment reduced the CPU-only training-preparation harness from 734 to 92 ms for SmolLM2, 387 to 52 ms for SmolVLA, and 521 to 88 ms for diffusion. Bounded alternative enumeration itself improves about 11–15% in these probes. Methods and limits are in docs/compiler-search.md; these are not GPU speedup claims.
